fix: enable to use secrets with special characters

mpatlasov opened the pull request:
If the password to SMB-server contained special characters (e.g. "foo,bar"), the mount failed. Now, when the password is passed to mount via "credentials=filename" option, then mount succeeds.

/kind bug

Fixes https://github.com/kubernetes-csi/csi-driver-smb/issues/507, https://github.com/kubernetes-csi/csi-driver-smb/issues/358, https://github.com/kubernetes-csi/csi-driver-smb/issues/573  

**Release note**:
```
none
```

andyzhangx reviewed (commented):
/ok-to-test

andyzhangx commented on pkg/smb/nodeserver.go:
```diff
@@ -231,7 +231,7 @@ func (d *Driver) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRe
 		}
 		if requireUsernamePwdOption && !useKerberosCache {
-			sensitiveMountOptions = []string{fmt.Sprintf("%s=%s,%s=%s", usernameField, username, passwordField, password)}
+			sensitiveMountOptions = []string{fmt.Sprintf("%s=%s", usernameField, username), fmt.Sprintf("%s=%s", passwordField, password)}
```
there could be problem when there is mount process concurrently, is it possible to detect whether there is special chars in secrets first, if yes, then mount with cred file? I think that would be safer since it won't break anything, thanks.

DanielZambelli replied:
sometimes I see password that contains " double quotes and other special chars. Currently we re-roll passwords to avoid these special chars. Using a temp config file would improve support.

andyzhangx commented:
btw, sanity tests all failed with this PR: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-csi_csi-driver-smb/961/pull-csi-driver-smb-sanity/1938798037342097408

johanot commented on pkg/smb/smb_common_linux.go:
```diff
@@ -20,12 +20,29 @@ limitations under the License.
 
 func Mount(m *mount.SafeFormatAndMount, source, target, fsType string, options, sensitiveMountOptions []string, _ string) error {
+	if len(sensitiveMountOptions) != 0 {
+		file, err := os.CreateTemp("/tmp/", "*.smb.credentials")
```
I would prefer this file on a ramfs so it wouldn't touch the disk at all. But... at the very minimum, the temp file must have restricted permissions, since /tmp is world-readable on most systems. This needs to be safe in the rare cases that the driver is not run in a container and/or with multiple processes sharing /tmp.

mpatlasov replied:
@johanot , the temp file already has restricted permissions (rw for the user only) because that's how os.CreateTemp() works. See its official docs here: https://pkg.go.dev/os#CreateTemp .

I intentionally used "/tmp" instead of ramfs, because the location of ramfs mount may vary depending on environment (at least, I don't know any path that would exist everywhere). But generally I fully agree -- not touching the disk would be better. And "defer os.Remove" will wipe it anyway.

mpatlasov commented:
/test pull-csi-driver-smb-e2e-windows-2022

mpatlasov commented:
@andyzhangx , I addressed your comments (and fixed test failures). Please take a look again.

andyzhangx commented on pkg/smb/smb_common_linux.go:
```diff
@@ -20,12 +20,38 @@ limitations under the License.
+// (see comments for ContainsSpecialCharacter() in pkg/smb/nodeserver.go).
+// NB: implementation relies on the format:
+// options := []string{fmt.Sprintf("%s=%s", usernameField, username), fmt.Sprintf("%s=%s", passwordField, password)}
+func NeedsCredentialsOption(options []string) bool {
```
@Copilot  maybe you could write a unit test now in the comments? let's see

mpatlasov replied:
Added new ut.

mpatlasov commented on pkg/smb/smb_common_linux.go:
```diff
@@ -20,12 +20,38 @@ limitations under the License.
+
 func Mount(m *mount.SafeFormatAndMount, source, target, fsType string, options, sensitiveMountOptions []string, _ string) error {
+	if NeedsCredentialsOption(sensitiveMountOptions) {
+		file, err := os.CreateTemp("/tmp/", "*.smb.credentials")
```
The file permission is already "rw" for the owner only, I verified that. So there must not be security risk there.

mpatlasov commented on pkg/smb/smb_common_linux.go:
```diff
@@ -20,12 +20,38 @@ limitations under the License.
+// NB: implementation relies on the format:
+// options := []string{fmt.Sprintf("%s=%s", usernameField, username), fmt.Sprintf("%s=%s", passwordField, password)}
+func NeedsCredentialsOption(options []string) bool {
+	return len(options) == 2 && strings.HasPrefix(options[1], "password=") && ContainsSpecialCharacter(options[1])
```
This is trade-off between code complexity and fragility. I resolved this trade-off toward keeping code simpler. I think it must be fine because we have exactly one place where options with len>1 can be generated. @andyzhangx , let me know if you disagree ("making this more flexible" can be easily done)

mpatlasov commented on pkg/smb/smb_common_linux.go:
```diff
@@ -20,12 +20,38 @@ limitations under the License.
+}
+
 func Mount(m *mount.SafeFormatAndMount, source, target, fsType string, options, sensitiveMountOptions []string, _ string) error {
+	if NeedsCredentialsOption(sensitiveMountOptions) {
```
This chunk (the implementation of func Mount) is already covered by TestNodeStageVolume from nodeserver_test.go. I'll just add another test case to use a password with special characters. This will lead execution inside this "if" clause.

andyzhangx commented on pkg/smb/smb_common_linux.go:
```diff
@@ -20,12 +20,38 @@ limitations under the License.
+				return err
+			}
+		}
+		file.Close()
```
@mpatlasov can you add following code in L44, others lgtm, thanks.

```
		defer func() {
			file.Close()
			os.Remove(file.Name())
		}()
```

mpatlasov replied:
OK, Done!

mpatlasov commented on pkg/smb/smb_common_linux.go:
```diff
@@ -20,12 +20,38 @@ limitations under the License.
+				return err
+			}
+		}
+		file.Close()
```
Let's ignore this suggestion. See this comment: https://github.com/kubernetes-csi/csi-driver-smb/pull/961/files#r2264676974

mpatlasov commented:
Hey @andyzhangx , I made three changes in [a new commit](https://github.com/kubernetes-csi/csi-driver-smb/pull/961/commits/a03aa0da9541061b81c1a427060341e2cb28a9d0):

 * Moved defer func to L44 (as you suggested [here](https://github.com/kubernetes-csi/csi-driver-smb/pull/961#discussion_r2264676974))
 * Added new test-case to existing TestNodeStageVolume to dive into Mount() "if NeedsCredentialsOption() {..."
 * Added new file `smb_common_linux_test.go` with tests for NeedsCredentialsOption()

mpatlasov commented on pkg/smb/smb_common_linux.go:
```diff
@@ -20,12 +20,43 @@ limitations under the License.
+				return err
+			}
+		}
+		if err := file.Close(); err != nil {
```
Done! (force-push of [second amended commit](https://github.com/kubernetes-csi/csi-driver-smb/pull/961/commits/5d48318a4f3c733f640ef91777edbb49fde4a285))

andyzhangx commented:
/retest

andyzhangx commented:
/retest

andyzhangx reviewed (approved):
/lgtm
